Add new examples and update existing ones (#260)
cirruslabs/cirrus-ci-docs: docs/examples.md

!!! info
    Please don't forget to setup [Remote Build Cache](#build-cache) for your Gradle project. Or at least [simple folder caching](#gradle-caching).

> ## Apache Maven

Reviewer: I don't think many people use the full name. You can simply use *Maven*.

> Here is an example `.cirrus.yml` that builds a Maven project (and runs basic tests on code to make sure there are no simple errors):

Reviewer: You can probably use similar working like this:

```suggestion
Official [Maven Docker images](https://hub.docker.com/_/maven/) can be used for building and testing Maven projects:
```

here to be more consistent.

> ```yaml

Reviewer: Add a new line before.

task:
  name: Cirrus CI
  container:

>     image: maven:3-jdk-8

Reviewer: JDK 11 is actually the default one. You can probably change the tag to `maven:3` which is `maven:3-jdk11`.

test_script: mvn compile -B
```

> If you want to use a different JDK version, you can change the `jdk-8` part to whatever version you want (e.g. `jdk-11`).

Reviewer: Don't think there is a need for it. Docker Hub page has all the information.

org.gradle.jvmargs=-Dfile.encoding=UTF-8
```

## JUnit

> Here is a `.cirrus.yml` that (once completed), runs and uploads JUnit results:

Reviewer: ```suggestion
Here is a `.cirrus.yml` that (once succeeded or failed), parses and uploads JUnit reports:
```

```yaml
junit_test_task:
  junit_script: <replace this comment with instructions to run the test suites>
  always:
    junit_result_artifacts:
      path: "**/test-results/**/*.xml"
      format: junit
```

If it is running on a pull request, annotations will also be displayed in-line.

## Maven

Official [Maven Docker images](https://hub.docker.com/_/maven/) can be used for building and testing Maven projects:

```yaml
task:
  name: Cirrus CI
  container:

>     image: maven:3-jdk-11

Reviewer: ```suggestion
    image: maven:latest
```

> test_script: mvn compile -B

Reviewer: ```suggestion
test_script: mvn test -B
```

Author: This doesn't work unless tests are specified

Reviewer: Are you sure? Back in the days it definitely worked. 🤔

Author: Just earlier today I set up a maven project, and running test returned no tests specified (but still exited with 0).  Compile checks for basic syntax errors.

Reviewer: Do you have any tests though?

Author: None set up.

Reviewer: That's maybe the issues. If you'll have any tests (which you should), `mvn test` will work.

org.gradle.jvmargs=-Dfile.encoding=UTF-8
```

## JUnit

Here is a `.cirrus.yml` that (once succeeded or failed), parses and uploads JUnit reports:

```yaml
junit_test_task:
  junit_script: <replace this comment with instructions to run the test suites>
  always:
    junit_result_artifacts:
      path: "**/test-results/**/*.xml"
      format: junit
```

If it is running on a pull request, annotations will also be displayed in-line.

## Maven

Official [Maven Docker images](https://hub.docker.com/_/maven/) can be used for building and testing Maven projects:

```yaml
task:
  name: Cirrus CI
  container:
    image: maven:latest
  maven_cache:
    folder: ~/.m2

>     fingerprint_script: cat pom.xml

Reviewer: It won't work for a project with multiple modules. I think you can simply remove `fingerprint_script`.
